### PR TITLE
[WIP] implemented Synology group/useradd modules

### LINKED ISSUE #42739: user management does not work under Synology
### Description of Issue/Question

Hi, on a Synology NAS I cannot manage users with `user` virtual module because no `useradd` command is present. While the installed version is pretty old, it does seem the same on develop branch.

I'd be happy to implement it, but I'm not sure on which choice between patch current module and writing a new would be better given that current useradd virtual match is quite large, matching [against kernel grain](https://github.com/saltstack/salt/blob/547ada9418442eaa9e87e380c411ac5e0d9a52e1/salt/modules/useradd.py#L40). What do you think ?

Thank you, regards

### Steps to Reproduce Issue
```shell
# salt-call --local user.add name=hello
[INFO    ] Executing command ['useradd', '-m', 'hello'] in directory '/root'
Error running 'user.add': Unable to run command ['useradd', '-m', 'hello'] with the context
 {...}, reason: [Errno 2] No such file or directory

```
### Versions Report
```
           Salt: 2015.5.3
         Python: 2.7.13 (default, Dec 18 2016, 21:23:50)
         Jinja2: 2.7.3
       M2Crypto: 0.25.1
 msgpack-python: 0.4.8
   msgpack-pure: Not Installed
       pycrypto: 2.6.1
        libnacl: Not Installed
         PyYAML: 3.12
          ioflo: Not Installed
          PyZMQ: 16.0.2
           RAET: Not Installed
            ZMQ: 4.1.6
           Mako: Not Installed
        Tornado: Not Installed
```

### PR DESCRIPTION
**Pull request by aogier**
### What does this PR do?
This PR implements a working group/useradd module for Synology NAS platforms

### What issues does this PR fix or reference?

closes #42739 

### Tests written?

extended existing ones

There's still a couple of minor issues to deal with, see this [issue comment](https://github.com/saltstack/salt/issues/42739#issuecomment-323703868) for reference.

### CODE REVIEW

**Comment by aogier** [comment]
Apart the pylint process which I completely missed (I'll work on that) I see that tests are failing because no `os_family` grain. Is there a grains subset know to be present on test nodes ? For the features I'm implementing I could choose on `os`, `os_family` and `osfullname` grains.
Thanks !

**Comment by aogier** [comment]
In console output I see a lot of errors on those grains:

```
 [salt.loader                                               :1687][ERROR   ] Exception raised when processing __virtual__ function for salt.loaded.int.module.grub_legacy. Module will not be loaded: 'os_family'
...
 [salt.loader                                               :1687][ERROR   ] Exception raised when processing __virtual__ function for salt.loaded.int.module.groupadd. Module will not be loaded: 'kernel'
```

It is maybe related somehow ? TIA

**Review by cachedout — COMMENTED** [review]
I think a better approach to do this might be to use a `provider` approach. Take a look at the `aix_group` module as one example, specifically the `__virtual__` function to see how provider overrides work and let me know what you think.

**Comment by aogier** [comment]
Hello, thank you for the feedback. When I begun working on this one I was wondering about the "right" way to implement it, either as a separate providers or in `G@kernel:Linux` existing provider (Synology runs on linux). The latter solution seemed better to me but a completely different, proprietary, userland requires lot of work and make clean code more messy either in implementation and in testcases logic.

So I'll definitely refactor everything into a standalone module leaving alone groupadd/useradd \<g\>, thank you for the help and suggestion and hope to get everything done as I badly need this for revamping gitlab formula and deploying runners :grimacing:.

File: tests/integration/modules/test_groupadd.py [hunk 1]
```diff
@@ -31,10 +31,11 @@ def setUp(self):
         self._no_group = self.__random_string()
         self._gid = 64989
         self._new_gid = 64998
-        os_grain = self.run_function('grains.item', ['kernel'])
-        if os_grain['kernel'] not in 'Linux':
```
Comment (aogier) [inline tests/integration/modules/test_groupadd.py]: existing bug

**Comment by cachedout** [comment]
Hi @aogier. Sounds good. Please just let me know when you feel like this is ready for another review. Thanks!

**Comment by aogier** [comment]
sure. Thank you !

**Comment by cachedout** [comment]
HI @aogier. Any updates here?

**Comment by cachedout** [comment]
@aogier I didn't hear back from you so I am closing this for the time being. When you are ready to come back to it please let me know and we can re-open this one or start on a new PR. Thanks.

**Pull request closed without merging**